open_ce/build_tree.py

```diff
@@ -108,93 +114,96 @@ def name(self):
-                                                                                         combined_config_files,
-                                                                                         variants)
+    def __eq__(self, other):
+        if not isinstance(other, DependencyNode):
+            return False
+        if self.build_command is not None and other.build_command is not None:
+            return self.packages == other.packages and self.build_command == other.build_command
+        return self.packages == other.packages
 
-        build_commands.append(BuildCommand(recipe=recipe.get('name', None),
-                                    repository=repository,
-                                    packages=packages,
-                                    version=version,
-                                    recipe_path=recipe_path,
-                                    runtime_package=runtime_package,
-                                    output_files=output_files,
-                                    python=variants['python'],
-                                    build_type=variants['build_type'],
-                                    mpi_type=variants['mpi_type'],
-                                    cudatoolkit=variants['cudatoolkit'],
-                                    run_dependencies=run_deps,
-                                    host_dependencies=host_deps,
-                                    build_dependencies=build_deps,
-                                    test_dependencies=test_deps,
-                                    channels=channels))
+def traverse_build_commands(build_tree, starting_nodes=None, return_node=False):
+    """
+    Generator function that goes through a list of BuildCommand's dependency tree.
+    """
+    if starting_nodes is not None:
+        false_start_node = "Starting Node"
+        new_graph = build_tree.copy()
+        new_graph.add_node(false_start_node)
+        for dep in starting_nodes:
+            new_graph.add_edge(false_start_node, dep)
+        generator = networkx.dfs_postorder_nodes(new_graph, false_start_node)
+    else:
+        generator = networkx.dfs_postorder_nodes(build_tree)
+    for current in generator:
+        if isinstance(current, DependencyNode):
+            if current.build_command is not None:
+                if return_node:
+                    yield current
+                else:
+                    yield current.build_command
+
+def get_independent_runtime_deps(tree, node):
+    """
+    This function gets all run dependencies of a node that don't depend on
+    any internal build commands.
+    """
+    deps = set()
+    if node.build_command is not None:
+        run_deps = {x for x in node.build_command.run_dependencies
+                                if utils.remove_version(x) not in map(utils.remove_version, node.packages)}
+        for run_dep in run_deps:
+            run_dep_node = {x for x in tree.successors(node)
```

Comment:
What is the pop for?

Reply:
I'm using the set builder notation to do a filter and since that returns a list I pop off the first element. I could probably do a find.
I looked and I could probably do this a little cleaner with the `next` keyword instead of pop.

Reply:
I changed this to use next.